### ISSUE TITLE
SSH authentication error from msqldump when pushing database (after successful file pushes)

### ISSUE THREAD

**Issue by ThinkByDesign** [welaika/wordmove#621]
So, to be clear, I'm new at wordmove. This is my second site with it, and the first one isn't quite working yet, but for different reasons.

**Describe the bug**

When wordmove gets to the remote dump stage, it generates the error noted below (SSH authentication). What's confusing is that it's just successfully conducted SSH connections to move all the files successfully. I'm also able to ssh directly from the command line, as well as through FileZilla.

When I login to the web server and look at the wp-content directory, I don't see a dump file there. HOWEVER, if I login to the web server (using the credential in the config file) and just run the mysqldump (substituting real values for the secrets omitted in the error), it works, and generates the dump.sql file.

I've also looked at the auth.log for possible clues. When I do a push with --all, I see several chunks like this:
Oct  6 02:22:16 web02 sshd[8898]: Accepted publickey for [correct user] from 99.164.33.88 port 57945 ssh2: [key stuff]
Oct  6 02:22:16 web02 sshd[8898]: pam_unix(sshd:session): session opened for user [correct user] by ([correct uid])
Oct  6 02:22:16 web02 systemd-logind[1422]: New session 82907 of user [correct user].
Oct  6 02:22:17 web02 sshd[8974]: Received disconnect from 99.164.33.88 port 57945:11: disconnected by user
Oct  6 02:22:17 web02 sshd[8974]: Disconnected from user [correct user] 99.164.33.88 port 57945
Oct  6 02:22:17 web02 sshd[8898]: pam_unix(sshd:session): session closed for user [correct user]
Oct  6 02:22:17 web02 systemd-logind[1422]: Removed session 82907.

but then when it tries to do the database thing, it just gets this:
Oct  6 02:22:22 web02 sshd[9364]: Connection closed by authenticating user [correct user - same as above] 99.164.33.88 port 57981 [preauth]

It's almost like it's using different credentials for that step? I do have my ssh config setup to filter by user and apply different keys (I have two logins to the same server with their own keys - both work correctly from the command line, but maybe this code is doing something different?)

Any help would be appreciated!!

**Wordmove command**
wordmove push -e staging -all

**Expected behavior**
Pushing all the files up to the remote staging server (which succeeded).
Pushing database up to the remote server (failed).

**movefile.yml**

```yml
global:
  sql_adapter: wpcli
  #sql_adapter: default #possible alternative if I run into issues
local:
  vhost: http://local.greenmeadowstx.com
  wordpress_path: /Users/kep/Sites/GreenMeadowsTX.com/wordpress # use an absolute path here
  database:
    name: wp_GreenMeadowsTXcom
    user: GreenMeadows
    password: "*************" # could be blank, so always use quotes around
    host: localhost
    mysqldump_options: '--socket /Applications/MAMP/tmp/mysql/mysql.sock'
    mysql_options: '--socket /Applications/MAMP/tmp/mysql/mysql.sock'

staging:
  vhost: https://greenmeadowstx.thinkbydesign.com
  wordpress_path: [absolute path to where wordpress is installed] # use an absolute path here
  
  database:
    name: GreenMeadowsTX_Wordpress
    user: [RemoteUserName]
    password: "[RemotePassword]"
    host: database.visible.ninja
 
  ssh:
    host: web02.visible.ninja
    # User, HostName, Port, and IdentityFile are all set in ~/.ssh/config
    user: [user with full access to the wordpress directory - key files are specified in ssh config and independently tested/working]
    #port: [port number] # Port is optional
    rsync_options: '--verbose --itemize-changes' # Additional rsync options, optional

  exclude:
    - '.git/'
    - '.gitignore'
    - '.gitmodules'
    - '.env'
    - 'node_modules/'
    - 'bin/'
    - 'tmp/*'
    - 'Gemfile*'
    - 'Movefile'
    - 'movefile'
    - 'movefile.yml'
    - 'movefile.yaml'
    - 'wp-config.php'
    - 'wp-content/*.sql.gz'
    - '*.orig'
```

**Exception/trace**

```
   remote | mysqldump --host=[secret] --user=wordpressuser --password=[secret] --result-file="[secret]/wp-content/dump.sql" GreenMeadowsTX_Wordpress
/Users/kep/.rvm/gems/ruby-2.3.0/gems/net-ssh-6.1.0/lib/net/ssh.rb:268:in `start': Authentication failed for user [REDACTED]@web02.visible.ninja (Net::SSH::AuthenticationFailed)
	from /Users/kep/.rvm/gems/ruby-2.3.0/gems/photocopier-1.4.0/lib/photocopier/ssh.rb:71:in `session'
	from /Users/kep/.rvm/gems/ruby-2.3.0/gems/photocopier-1.4.0/lib/photocopier/ssh.rb:44:in `exec!'
	from /Users/kep/.rvm/gems/ruby-2.3.0/gems/wordmove-5.2.2/lib/wordmove/deployer/ssh.rb:66:in `remote_run'
	from /Users/kep/.rvm/gems/ruby-2.3.0/gems/wordmove-5.2.2/lib/wordmove/deployer/ssh.rb:79:in `download_remote_db'
	from /Users/kep/.rvm/gems/ruby-2.3.0/gems/wordmove-5.2.2/lib/wordmove/deployer/ssh/wpcli_sql_adapter.rb:6:in `backup_remote_db!'
	from /Users/kep/.rvm/gems/ruby-2.3.0/gems/wordmove-5.2.2/lib/wordmove/deployer/ssh.rb:37:in `push_db'
	from /Users/kep/.rvm/gems/ruby-2.3.0/gems/wordmove-5.2.2/lib/wordmove/cli.rb:116:in `block in push'
	from /Users/kep/.rvm/gems/ruby-2.3.0/gems/wordmove-5.2.2/lib/wordmove/cli.rb:40:in `block in handle_options'
	from /Users/kep/.rvm/gems/ruby-2.3.0/gems/wordmove-5.2.2/lib/wordmove/cli.rb:39:in `each'
	from /Users/kep/.rvm/gems/ruby-2.3.0/gems/wordmove-5.2.2/lib/wordmove/cli.rb:39:in `handle_options'
	from /Users/kep/.rvm/gems/ruby-2.3.0/gems/wordmove-5.2.2/lib/wordmove/cli.rb:115:in `push'
	from /Users/kep/.rvm/gems/ruby-2.3.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	from /Users/kep/.rvm/gems/ruby-2.3.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	from /Users/kep/.rvm/gems/ruby-2.3.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	from /Users/kep/.rvm/gems/ruby-2.3.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
	from /Users/kep/.rvm/gems/ruby-2.3.0/gems/wordmove-5.2.2/exe/wordmove:6:in `<top (required)>'
	from /Users/kep/.rvm/gems/ruby-2.3.0/bin/wordmove:23:in `load'
	from /Users/kep/.rvm/gems/ruby-2.3.0/bin/wordmove:23:in `<main>'
```

**Environment (please complete the following information):**
- OS: MacOS 11.6
- Ruby: 2.3.0
- Wordmove: 5.2.2

**Doctor**
Reports all green except for 
    ❌  error | [/database/mysql_options] key 'mysql_options:' is undefined.

**Comment by ThinkByDesign** [comment]
(Update: just saw your reply while typing this. I'll try that and see what happens.)

OK. That actually matches my original assumption - that from my local machine, there's an ssh connection to the web server, and then the web server connects to the database server just like wordpress does. That also makes sense based on how the config file is setup, etc.

The only clue not matching that scenario is that the database pull experiences an ssh authentication failure, which APPEARS to be a connection to the database server as root. (It's possible that I'm seeing coincidental hacking attempts as well, though, and jumping to an inaccurate conclusion.)

**Comment by pioneerskies** [comment]
The fact is that if you have `ssh.user` it will superseed the one in `~/.ssh/config`. Now it's impossible for me to read deeply how the SSH ruby's implementation will behave in such a scenario, but if you have some directive like "match user xxx" then you're risking to not match it because you're passing another one within `movefile.yml`.

Think of a situation where you have

```
Host foo_host
  User foo_user
  Hostname 1.1.1.1

Match User foo_user
  Do something with your keys
```

and you're calling SSH like

```
ssh -l bar_user foo_host
```

> this could be something similar to what's happening when Wordmove calls the SSH routines

then I fear this would blow up your `Match` stanza. Not sure this is your case, but I'd check into it in your place.

**Comment by pioneerskies** [comment]
I've read the part of `Net::SSH` (ruby's core lib handling SSH connections) which handles SSH's config files: `Net::SSH::Config` https://github.com/net-ssh/net-ssh/blob/a859e03c5845db1579bceb56482636abd96b2d3e/lib/net/ssh/config.rb

> NOTE: If you're not a rubist do not waste time on code implementation: it looks hard for me, whom write quite a bit of ruby, and it's a mess of regular expressions. But comments are interesting for everyone.

From what I understand the only supported `Match`es statements are `all` and `host`. So I'd confirm `Match user` is not parsed and no warning is thrown to STDOUT.

So yes, your assumption was just on point: Wordmove doesn't (cannot) support `Match user` neither. This could go into some part of documentation.

A polite way to handle such situation would be to use two different `Host` instead of a `Match` on the user, e.g.:

```
Host foouser_websrv
  User foouser
  Hostname 1.1.1.1
  IdentityFile /some/key

Host baruser_websrv
  User baruser
  Hostname 1.1.1.1
  IdentityFile /some/other/key
```

I've retagged and assigned the ticket to me in order to take the time to think where to put a note about this fact in the docs.

Thank you very much for your detailed troubleshooting and for the time invested.

**Comment by pioneerskies** [comment]
Going to close as there are not more actions to go for

**Comment by humblecoder** [comment]
Apologies for resurrecting a "dead" thread, but `﻿﻿Net::SSH::AuthenticationFailed` is still happening for me.

Currently running:  MacOS 12.4 on M1 Max

I've tried every single version of ruby from 2.6.0 to 3.1.1, along with each relevant Wordmove gem.  None work, and the majority fail with `Net::SSH::AuthenticationFailed`.  I did this on both the host machine as well as a Vagrant Linux VM.  It would be nice, at the very least, to get a bit more verbose output regarding what failed.

I am able to manually SSH to the server via ssh keys from both the host and guest machines.  Wordmove (or more specifically net-ssh gem doesn't seem to care).

Any additional thoughts?  🤔 

**UPDATE**
Turns out [Issue 586](https://github.com/welaika/wordmove/issues/586) was the key (pun intended) to my issue.  😖
